# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [2]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [3]:
def inspect_data(_data, data_name):
    print(f"\nSummary information for {data_name} table")
    print("-"*40)
    print(_data.head(5))
    print(_data.info())
    print(_data.describe())
    

inspect_data(inventories, "inventories")
inspect_data(inventory_sets, "inventory sets")
inspect_data(sets, "sets")
inspect_data(themes, "themes")
inspect_data(inventory_minifigs, "inventory minifigs")
inspect_data(minifigs, "minifigs")
inspect_data(inventory_parts, "inventory parts")
inspect_data(part_categories, "part_categories")
inspect_data(parts, "parts")
inspect_data(colors, "colors")
inspect_data(part_relationships, "part_relationships")
inspect_data(elements, "elements")


Summary information for inventories table
----------------------------------------
   id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB
None
                  id       version
count   31587.000000  31587.000000
mean    43538.531389      1.077405
std     32458.542329      0.515993
min         1.000000      1.000000
25%     12193.500000      1.000000
50%     51076.000000      1.000000
75%     62254.500000      1.000000
max    122472.000000     11.000000

Summary information for inventory sets table
----------------------------------------
   inventory_id  set_num  

None
                id
count   190.000000
mean    479.394737
std     820.735479
min      -1.000000
25%      61.250000
50%     182.500000
75%    1015.750000
max    9999.000000

Summary information for part_relationships table
----------------------------------------
  rel_type child_part_num parent_part_num
0        P    87079pr9974           87079
1        P     3960pr9971            3960
2        R    98653pr0003     98086pr0003
3        R    98653pr0003    98088pat0003
4        R    98653pr0003    98089pat0003
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24508 entries, 0 to 24507
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   rel_type         24508 non-null  object
 1   child_part_num   24508 non-null  object
 2   parent_part_num  24508 non-null  object
dtypes: object(3)
memory usage: 574.5+ KB
None
       rel_type child_part_num parent_part_num
count     24508          24508           24508
unique 

## Assignment
Answer all the questions below using the available data. There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [4]:
def data_count(dataset):
    return len(dataset)

data_count(sets)

18576

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [5]:
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [6]:
Q2 = int(sets['num_parts'].mean())
Q2

160

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [7]:
Q3 = int(sets['num_parts'].median())
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [8]:
Q4 = sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [9]:
sets.loc[sets['num_parts'] == Q4]

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [10]:
def select_set_row(set_num: str):
    return sets.loc[sets['set_num'] == set_num]

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [11]:
def select_set_numparts(num_part: int):
    sets_copy = sets.copy()
    query = sets_copy.loc[sets_copy['num_parts'] == num_part].sort_values('year')
    return query

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [12]:
print(len(select_set_numparts(0)))
query = select_set_numparts(0)
query

3146


set_num                                       name  year  theme_id  \
11061   700.1-2             Extra-Large Gift Set (Mursten)  1953       366   
17563  b55dk-01       Lego Mursten - System i Leg Byggebog  1955       497   
17564  b56de-01                            System im Spiel  1956       497   
11072   700.2-2                               Gift Package  1957       366   
11078  700.3A-2                               Gift Package  1957       372   
...         ...                                        ...   ...       ...   
12334   71402-0       Character Pack Series 4 - Random Bag  2022       690   
12337  71402-11  Character Pack Series 4 - Complete Series  2022       690   
13421   76203-1                        Iron Man Mech Armor  2022       696   
13422   76204-1                   Black Panther Mech Armor  2022       696   
5293    40553-1                          Woody and Bo Peep  2022       610   

       num_parts  
11061          0  
17563          0  
17564          0  
11072          0  
11078          0  
...          ...  
12334          0  
12337          0  
13421          0  
13422          0  
5293           0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [13]:
Q8A = query.year.iloc[0]
Q8A

1953

In [14]:
Q8B = query.year.iloc[-1]
Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [15]:
def select_set_year(year : int):
    sets_copy = sets.copy()
    query = sets_copy.loc[sets_copy['year'] == year].sort_values('set_num')
    return query

Q9A = select_set_year(1953)
Q9A

set_num                                               name  year  \
11061    700.1-2                     Extra-Large Gift Set (Mursten)  1953   
11066  700.1.4-1                            Individual 1 x 2 Bricks  1953   
11077  700.3.4-1                            Individual 2 x 3 Bricks  1953   
11093    700.F-1  Automatic Binding Bricks - Small Brick Set (Le...  1953   

       theme_id  num_parts  
11061       366          0  
11066       371          1  
11077       371          1  
11093       371         48

In [16]:
Q9A = select_set_year(2022)
Q9A

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [17]:
Q10 = int(sets['year'].median())
Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [18]:
sets_copy = sets.copy()

Q11 = int(sets_copy.groupby('year')['year'].count().mean())
Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [19]:
sets_themes = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id', how = 'outer')
sets_themes.rename(columns = {'name_x':'name_sets', 'name_y':'name_themes'},
                  inplace = True)

# print(sets_themes.head())

def theme_by_year(year : int):
    theme_names = sets_themes.loc[sets_themes['year'] == year]
    theme_names = theme_names[['id', 'name_themes']]
    theme_names.reset_index(inplace = True)
    theme_names.drop('index', axis = 1, inplace = True)
    return theme_names

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [20]:
theme_by_year(1960)

id   name_themes
0  497         Books
1  371  Supplemental
2  513       Classic

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [21]:
Q13 = sets_themes.groupby('name_themes')['name_sets'].count()
Q13 = pd.DataFrame(Q13)
most_set_num = Q13['name_sets'].max()
Q13 = Q13.loc[Q13['name_sets'] == most_set_num]
Q13

name_sets
name_themes           
Gear              1904

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [22]:
def theme_by_name(theme_name):
    sets_ = sets_themes.loc[sets_themes['name_themes'] == theme_name]
    sets_ = sets_[['set_num', 'name_sets', 'year', 'theme_id', 'name_themes']]
    sets_.reset_index(inplace = True)
    sets_.drop('index', axis = 1, inplace = True)
    return sets_

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [23]:
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  theme_id name_themes  
0     2008       501        Gear  
1     2021       501        Gear  
2     2021       501        Gear  
3     2021       501        Gear  
4     2021       501        Gear  
...    ...       ...         ...  
1899  1997       501        Gear  
1900  2007       501        Gear  
1901  2008       501        Gear  
1902  2021       501        Gear  
1903  2020       501        Gear  

[1904 rows x 5 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [24]:
def theme_by_setnum(set_num: str):
    sets_ = sets_themes.loc[sets_themes['set_num'] == set_num]
    return sets_['name_themes'].unique()

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [25]:
theme_by_setnum('001-1')

array(['Technic'], dtype=object)

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [26]:
Millennium_Falcon = sets_themes.loc[sets_themes['name_sets'] == 'Millennium Falcon']
max_num_parts = Millennium_Falcon['num_parts'].max()
Millennium_Falcon.loc[Millennium_Falcon['num_parts'] == max_num_parts]

set_num          name_sets  year  theme_id  num_parts   id  \
3199  75192-1  Millennium Falcon  2017       171       7541  171   

                    name_themes  parent_id  
3199  Ultimate Collector Series      158.0

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [27]:
inventories_minifig = pd.merge(inventories, inventory_minifigs, left_on = 'id', right_on = 'inventory_id', how = 'inner')
inventories_minifig.head()
minifig = inventories_minifig.loc[inventories_minifig['set_num'] == '75192-1']
print(minifig.head())
Q17 = len(minifig)
Q17

          id  version  set_num  inventory_id     fig_num  quantity
11492  19670        1  75192-1         19670  fig-001714         1
11493  19670        1  75192-1         19670  fig-001810         1
11494  19670        1  75192-1         19670  fig-002057         1
11495  19670        1  75192-1         19670  fig-002514         1
11496  19670        1  75192-1         19670  fig-002544         1


8

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [28]:
inventories_sets = pd.merge(inventory_sets, sets, on = 'set_num', how = 'outer')
inventories_sets.rename(columns = {'name':'name_sets', 'num_parts':'sets_num_parts', 
                                       'quantity':'sets_quantity'},
                                   inplace = True)
inventories_minifigs = pd.merge(inventory_minifigs, minifigs, on = 'fig_num', how ='outer')
inventories_minifigs.rename(columns = {'name':'fig_name', 'num_parts':'fig_num_parts', 
                                       'quantity':'fig_quantity'},
                                       inplace = True)
sets_minifigs = pd.merge(inventories_minifigs, inventories_sets , on = 'inventory_id', how = 'outer')
inventory_sets_minifigs = pd.merge(sets_minifigs, inventories , on = 'set_num', how = 'outer').drop_duplicates()
inventory_sets_minifigs.head()

inventory_id     fig_num  fig_quantity  \
0           3.0  fig-001549           1.0   
1           4.0  fig-000764           1.0   
2       14589.0  fig-000764           1.0   
3          19.0  fig-000555           1.0   
4          25.0  fig-000574           1.0   

                                            fig_name  fig_num_parts set_num  \
0                 Emma - Lavender Top, Magenta Skirt            4.0     NaN   
1                    Danny Longlegs / Corporal Steel            4.0     NaN   
2                    Danny Longlegs / Corporal Steel            4.0     NaN   
3                               Coca-Cola Defender 2            4.0     NaN   
4  Minnie Mouse with Dark Pink with White Spots D...            5.0     NaN   

   sets_quantity name_sets  year  theme_id  sets_num_parts  id  version  
0            NaN       NaN   NaN       NaN             NaN NaN      NaN  
1            NaN       NaN   NaN       NaN             NaN NaN      NaN  
2            NaN       NaN   NaN       NaN             NaN NaN      NaN  
3            NaN       NaN   NaN       NaN             NaN NaN      NaN  
4            NaN       NaN   NaN       NaN             NaN NaN      NaN

In [29]:
def get_minifigs(set_num : str, version: int):
    table = inventory_sets_minifigs.loc[(inventory_sets_minifigs['set_num'] == set_num) & (inventory_sets_minifigs['version'] == version)]
    table = table[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']]
    table.reset_index(inplace = True)
    table.drop('index', axis = 1, inplace = True)
    return table
    

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [30]:
Q19 = get_minifigs('75192-1', 1)
Q19

set_num  version          name_sets     fig_num  \
0    75192-1      1.0  Millennium Falcon  fig-001505   
1    75192-1      1.0  Millennium Falcon  fig-001806   
2    75192-1      1.0  Millennium Falcon  fig-001877   
3    75192-1      1.0  Millennium Falcon  fig-001881   
4    75192-1      1.0  Millennium Falcon  fig-001884   
..       ...      ...                ...         ...   
131  75192-1      1.0  Millennium Falcon  fig-012221   
132  75192-1      1.0  Millennium Falcon  fig-012222   
133  75192-1      1.0  Millennium Falcon  fig-012223   
134  75192-1      1.0  Millennium Falcon  fig-012224   
135  75192-1      1.0  Millennium Falcon  fig-012225   

                                  fig_name  
0      Lime Torso, Lime Legs, Black Helmet  
1                              Cheep Cheep  
2                  Dwayne - Possessed Form  
3           Captain Jonas - Possessed Form  
4             Mr. Branson - Possessed Form  
..                                     ...  
131                                Bob-omb  
132                   Boomer (Banzai) Bill  
133                            Bullet Bill  
134                           Buzzy Beetle  
135  Unikitty - Cheerykitty (Cheery Kitty)  

[136 rows x 5 columns]

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [31]:
get_minifigs('10217-1', 1)

set_num  version     name_sets     fig_num  \
0    10217-1      1.0  Diagon Alley  fig-001505   
1    10217-1      1.0  Diagon Alley  fig-001806   
2    10217-1      1.0  Diagon Alley  fig-001877   
3    10217-1      1.0  Diagon Alley  fig-001881   
4    10217-1      1.0  Diagon Alley  fig-001884   
..       ...      ...           ...         ...   
131  10217-1      1.0  Diagon Alley  fig-012221   
132  10217-1      1.0  Diagon Alley  fig-012222   
133  10217-1      1.0  Diagon Alley  fig-012223   
134  10217-1      1.0  Diagon Alley  fig-012224   
135  10217-1      1.0  Diagon Alley  fig-012225   

                                  fig_name  
0      Lime Torso, Lime Legs, Black Helmet  
1                              Cheep Cheep  
2                  Dwayne - Possessed Form  
3           Captain Jonas - Possessed Form  
4             Mr. Branson - Possessed Form  
..                                     ...  
131                                Bob-omb  
132                   Boomer (Banzai) Bill  
133                            Bullet Bill  
134                           Buzzy Beetle  
135  Unikitty - Cheerykitty (Cheery Kitty)  

[136 rows x 5 columns]

In [32]:
get_minifigs('10217-1', 2)

set_num  version     name_sets     fig_num  \
0    10217-1      2.0  Diagon Alley  fig-001505   
1    10217-1      2.0  Diagon Alley  fig-001806   
2    10217-1      2.0  Diagon Alley  fig-001877   
3    10217-1      2.0  Diagon Alley  fig-001881   
4    10217-1      2.0  Diagon Alley  fig-001884   
..       ...      ...           ...         ...   
131  10217-1      2.0  Diagon Alley  fig-012221   
132  10217-1      2.0  Diagon Alley  fig-012222   
133  10217-1      2.0  Diagon Alley  fig-012223   
134  10217-1      2.0  Diagon Alley  fig-012224   
135  10217-1      2.0  Diagon Alley  fig-012225   

                                  fig_name  
0      Lime Torso, Lime Legs, Black Helmet  
1                              Cheep Cheep  
2                  Dwayne - Possessed Form  
3           Captain Jonas - Possessed Form  
4             Mr. Branson - Possessed Form  
..                                     ...  
131                                Bob-omb  
132                   Boomer (Banzai) Bill  
133                            Bullet Bill  
134                           Buzzy Beetle  
135  Unikitty - Cheerykitty (Cheery Kitty)  

[136 rows x 5 columns]

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [33]:
def minfigs_from_themes(theme_id: int):
    table = inventory_sets_minifigs.loc[inventory_sets_minifigs['theme_id'] == theme_id]
    table = table[['fig_num', 'fig_name']]
    table.sort_values('fig_num')
    table.reset_index(inplace = True)
    table.drop('index', axis = 1, inplace = True)
    return table

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [34]:
minfigs_from_themes(171)

fig_num                               fig_name
0     fig-001505    Lime Torso, Lime Legs, Black Helmet
1     fig-001806                            Cheep Cheep
2     fig-001877                Dwayne - Possessed Form
3     fig-001881         Captain Jonas - Possessed Form
4     fig-001884           Mr. Branson - Possessed Form
...          ...                                    ...
5980  fig-012222                   Boomer (Banzai) Bill
5981  fig-012223                            Bullet Bill
5982  fig-012224                           Buzzy Beetle
5983  fig-012225  Unikitty - Cheerykitty (Cheery Kitty)
5984         NaN                                    NaN

[5985 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [35]:
def sets_from_minifig(fig_num :str):
    table = inventory_sets_minifigs.loc[inventory_sets_minifigs['fig_num'] == fig_num]
    table = table[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']]
    table.sort_values(['set_num', 'version'])
    table.reset_index(inplace = True)
    table.drop('index', axis = 1, inplace = True)
    return table

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [36]:
sets_from_minifig('fig-001039')

set_num  version name_sets     fig_num  \
0     NaN      NaN       NaN  fig-001039   
1     NaN      NaN       NaN  fig-001039   
2     NaN      NaN       NaN  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [46]:
inventories_parts = pd.merge(inventory_parts, parts, on = 'part_num', how = 'inner')
inventories_parts.rename(columns = {'name':'name_part', },
                                   inplace = True)
sets_themes = pd.merge(sets, themes, left_on = 'theme_id', right_on = 'id', how = 'inner')
sets_themes.rename(columns = {'name_x':'name_sets', 'name_y':'name_themes'},
                  inplace = True)

sets_themes_inventories = pd.merge(sets_themes, inventories , on = 'set_num', how = 'inner')
sets_themes_inventories.rename(columns = {'id_x':'inventory_id'},
                  inplace = True)

part_inventories = pd.merge(inventories_parts, inventories , left_on = 'inventory_id', 
                            right_on = 'id', how = 'inner')

In [49]:
sets_themes_parts_inventories = pd.concat([sets_themes_inventories, 
                                         part_inventories]).drop_duplicates()
sets_themes_parts_inventories.head(5)

set_num                       name_sets    year  theme_id  num_parts  \
0   001-1                           Gears  1965.0       1.0       43.0   
1   002-1  4.5V Samsonite Gears Motor Set  1965.0       1.0        3.0   
2  1030-1  TECHNIC I: Simple Machines Set  1985.0       1.0      211.0   
3  1038-1              ERBIE the Robo-Car  1985.0       1.0      120.0   
4  1039-1            Manual Control Set 1  1986.0       1.0       39.0   

   inventory_id name_themes  parent_id     id_y  version part_num  color_id  \
0             1     Technic        NaN  24696.0        1      NaN       NaN   
1             1     Technic        NaN  24701.0        1      NaN       NaN   
2             1     Technic        NaN  15686.0        1      NaN       NaN   
3             1     Technic        NaN  13420.0        1      NaN       NaN   
4             1     Technic        NaN  10042.0        1      NaN       NaN   

   quantity is_spare name_part  part_cat_id part_material  id  
0       NaN      NaN       NaN          NaN           NaN NaN  
1       NaN      NaN       NaN          NaN           NaN NaN  
2       NaN      NaN       NaN          NaN           NaN NaN  
3       NaN      NaN       NaN          NaN           NaN NaN  
4       NaN      NaN       NaN          NaN           NaN NaN

In [ ]:
# not complete
def set_parts(set_num : str, version : int):
    table = inventory_sets_minifigs.loc[inventory_sets_minifigs['fig_num'] == fig_num]
    table = table[['set_num', 'version', 'name_sets', 
                   'year', 'name_themes', 'part_num',
                   'name_colr', 'quantity', 'is_spare']]
    table.sort_values(['set_num', 'version'])
    table.reset_index(inplace = True)
    table.drop('index', axis = 1, inplace = True)
    return table

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [ ]:
### CODE CHECK ###

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [ ]:
### ENTER CODE HERE ###

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [ ]:
### ENTER CODE HERE ###

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [ ]:
### ENTER CODE HERE ###

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
